### PR TITLE
Add support for custom metadata serializers to QPY load() and dump()

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a new argument to load(), metadata_serializer, and
dump(), metadata_deserializer, which is used to specify custom
serialization functions to use for serializing the python dictionary for
each circuit's metadata attribute. By default QPY expects the serialized
metadata payload to be encoded as a utf8 json string. But since there
are no user constraints on the metadata dictionary's contents a simple
json.dump() of the circuit's metadata attribute may not be feasible. To
address this issue these new arguments are added to enable users to
specify a custom function to encode the dictionary as binary data
however is required for their circuit's metadata and then read that
custom format as required when deserializing.

The QPY format specification does not change here and still explicitly
lists a UTF8 encoded JSON string as the contents. A fixed format is
needed in the format specification for potential interoperability for
other tools generating QPY data (none exist currently to my knowledge
but nothing precludes it which is why the format is publicly
documented). These new arguments are specific to only the qiskit
implementation and should only be used when generating qpy files and
loading them via qiskit.

### Details and comments

### CODE REVIEW

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+features:
+  - |
+    Added a new kwarg, ``metadata_serializer`` to the
+    :func:`.qpy_serialization.dump` function for specifying a custom
+    serialization function for use when serializing the
+    :attr`.QuantumCircuit.metadata` attribute. By default the
+    :func:`~qiskit.circuit.qpy_serialization.dump` function will attempt to
```
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:8]: for laziness: you can do both `~` and `.`, so `` :func:`~.qpy_serialization.dump` `` does the lookup with the `.` semantics, and displays it using the `~` truncation semantics (so it just shows `dump`).  No change needed here, though.

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    JSON serialize the contents of :attr`.QuantumCircuit.metadata` for each
+    circuit, but since :attr`.QuantumCircuit.metadata` can contain any python
+    dictionary even those with contents not JSON serializable by the default
+    encoder will lead to circuits that can't be serialized. The new
```
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:12]:
```suggestion
    dictionary, even those with contents not JSON serializable by the default
    encoder, will lead to circuits that can't be serialized. The new
```

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    that will be passed the :attr`.QuantumCircuit.metadata` dictionary and
+    will return a ``bytes`` object that represents the dictionary. This can
+    then be paired with the new ``metadata_deserializer`` argument of the
+    :func:`.qpy_serialization.load` function to load whatever the output is.
+    (If ``metadata_deserializer`` is not specified on
+    :func:`~qiskit.circuit.qpy_serialization.load` then the metadata payload
+    will be treated as a UTF8 encoded JSON string).
+  - |
+    Added a new kwarg, ``metadata_deserializer`` to the
+    :func:`.qpy_serialization.load` function for specifying a custom
```
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:23]: It seems a little repetitive to describe the changes to `dump` and `load` separately.

**Review by jakelishman — COMMENTED** [review]
As a raw implementation I don't see any problems, but I'm a bit concerned about breaking the binary format guarantees.  At the very least, the module documentation needs updating to mention that the `METADATA` field might not be utf8-encoded JSON data, but given that it's really an _arbitrary_ injected dependency, I wonder if it's best to put a new field into the metadata header in a new QPY format, just to indicate whether it needs a custom dependency.  I'm not 100% sold on that, because it feels like it couldn't fully specify the data anyway if the serialiser and deserialiser are speaking valid JSON, but with special rules for custom objects, but I'm most concerned with the case that the data isn't valid JSON at all.

If we don't add a header field (and even if we do), it might be good to try/catch the deserialisation in order to provide a better error message if we can't read the data (e.g. "did you remember to pass your custom deserialiser?")

**Comment by mtreinish** [comment]
> As a raw implementation I don't see any problems, but I'm a bit concerned about breaking the binary format guarantees. At the very least, the module documentation needs updating to mention that the `METADATA` field might not be utf8-encoded JSON data, but given that it's really an _arbitrary_ injected dependency, I wonder if it's best to put a new field into the metadata header in a new QPY format, just to indicate whether it needs a custom dependency. I'm not 100% sold on that, because it feels like it couldn't fully specify the data anyway if the serialiser and deserialiser are speaking valid JSON, but with special rules for custom objects, but I'm most concerned with the case that the data isn't valid JSON at all.
> 

Yeah, I was worried about that. I figured it would be good to just support that as an out of specification feature and if a user returns non-utf8 json bytes (like I did in the unittest) it's on them as being out of spec. That being said the other idea I had was to change the argument type to be an `JSONDecoder` and `JSONEncoder` subclass instead and then change the call to just set that on `json.dump()` and `json.load()` that way it will always be json.

**Comment by jakelishman** [comment]
Out-of-spec features have a tendency to morph into features that we actually have to support.  For safety's sake, I think I prefer _requiring_ the data to be JSON with your idea of passing a `JSONEncoder`/`JSONDecoder` instance - it's easier for us to relax the requirement in the future than to tighten it.

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    :attr`.QuantumCircuit.metadata` attribute and a dual kwarg
+    ``metadata_deserializer`` to the :func:`.qpy_serialization.load` function
+    for specifying a ``JSONDecoder`` subclass. By default the
+    :func:`~qiskit.circuit.qpy_serialization.dump` and
+    :func:`~qiskit.circuit.qpy_serialization.load` functions will attempt to
+    JSON serialize and deserialize with the stdlib default json encoder and
+    decoder. The contents of :attr`.QuantumCircuit.metadata` for each circuit,
+    but since :attr`.QuantumCircuit.metadata` can contain any Python
+    dictionary even those with contents not JSON serializable by the default
+    encoder will lead to circuits that can't be serialized. The new
```
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:16]: Something's off with this sentence - I'm not quite sure what it's saying.

**Review by jakelishman — COMMENTED** [review]
This looks stricter and safer now - I guess it promotes slightly less efficient serialisation, but being stricter about what's allowed in the format feels more reliable, since nothing will crash if you try and load an unknown QPY file now.

Just one comment about the release note.

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+            def default(self, o):  # pylint: disable=invalid-name
+                if isinstance(o, CustomObject):
+                    return {"__type__": "Custom", "value": o.string}
+                return json.JSONEncoder.default(self, o)
+
+        class CustomDeserializer(json.JSONDecoder):
+            """Custom json decoder to handle CustomObject."""
+
+            def object_hook(self, o):  # pylint: disable=invalid-name,method-hidden
```
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:55]: Do we need the pylint markers in the release note?  I didn't think we'd be checking those blocks.  If not, possibly best to get rid of them to make things cleaner for readers.

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            def __init__(self, string):
+                self.string = string
+
+            def __eq__(self, other):
+                return self.string == other.string
+
+        class CustomSerializer(json.JSONEncoder):
+            """Custom json encoder to handle CustomObject."""
+
+            def default(self, o):  # pylint: disable=invalid-name
```
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:47]:
```suggestion
            def default(self, o):
```

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml (hunk 5 above)
Comment (jakelishman) [inline releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:55]:
```suggestion
            def object_hook(self, o):
```

**Review by jakelishman — APPROVED** [review]
Looks good to me, one minor comment aside.

File: releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml (hunk 5 above)
Comment (mtreinish) [reply releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml:55]: Ah good catch, no we don't. I just copied this from the unittests verbatim Let's remove them

**Review by jakelishman — APPROVED** [review]

**Pull request merged**